presto-spi/src/main/java/io/prestosql/spi/type/SqlTimestamp.java

```diff
@@ -103,9 +165,33 @@ public String toString()
                 .map(ZoneId::of)
                 .orElse(ZoneOffset.UTC);
 
-        return Instant.ofEpochMilli(millis)
-                .atZone(zoneId)
-                .format(JSON_FORMATTER);
+        return formatTimestamp(precision, epochMicros, picosOfMicros, zoneId);
+    }
+
+    public static String formatTimestamp(int precision, long epochMicros, int picosOfMicro, ZoneId zoneId)
+    {
+        Instant instant = Instant.ofEpochSecond(Math.floorDiv(epochMicros, MICROSECONDS_PER_SECOND));
+        LocalDateTime dateTime = LocalDateTime.ofInstant(instant, zoneId);
+
+        StringBuilder builder = new StringBuilder();
+        builder.append(TIMESTAMP_FORMATTER.format(dateTime));
+        if (precision > 0) {
+            long picoFraction = ((long) floorMod(epochMicros, MICROSECONDS_PER_SECOND)) * PICOSECONDS_PER_MICROSECOND + picosOfMicro;
+            long scaledFraction = roundDiv(picoFraction, POWERS_OF_TEN[MAX_PRECISION - precision]);
+            builder.append(".");
+            builder.append(format("%0" + precision + "d", scaledFraction));
+        }
```

Comment:
For precision=0 the second fraction is truncated.
For precision>0, the value is rounded. 

See TODO here (& consider adding as a test case):

```java
@Test
public void testFormatTimestamp()
{
    assertEquals(formatTimestamp(0, -1_000_000, 123456, UTC), "1969-12-31 23:59:59");
    assertEquals(formatTimestamp(0, -1_000_000, 654321, UTC), "1969-12-31 23:59:59"); // TODO rounding
    assertEquals(formatTimestamp(0, 1_000_000, 123456, UTC), "1970-01-01 00:00:01");
    assertEquals(formatTimestamp(0, 1_000_000, 654321, UTC), "1970-01-01 00:00:01"); // TODO rounding

    assertEquals(formatTimestamp(9, -1_000_000, 0, UTC), "1969-12-31 23:59:59.000000000");
    assertEquals(formatTimestamp(9, -1_000_000, 123456, UTC), "1969-12-31 23:59:59.000000123");
    assertEquals(formatTimestamp(9, 0, 0, UTC), "1970-01-01 00:00:00.000000000");
    assertEquals(formatTimestamp(9, 0, 123456, UTC), "1970-01-01 00:00:00.000000123");
    assertEquals(formatTimestamp(9, 1_000_000, 123456, UTC), "1970-01-01 00:00:01.000000123");

    assertEquals(formatTimestamp(3, 1_234_567, 890123, UTC), "1970-01-01 00:00:01.235");
    assertEquals(formatTimestamp(6, 1_234_567, 890123, UTC), "1970-01-01 00:00:01.234568"); // rounding
    assertEquals(formatTimestamp(9, 1_234_567, 890123, UTC), "1970-01-01 00:00:01.234567890");
    assertEquals(formatTimestamp(12, 1_234_567, 890123, UTC), "1970-01-01 00:00:01.234567890123");

    // TODO cover other zones
}
```

Reply:
Good catch!

Reply:
Actually, no, it doesn't matter. It turns out that the values are already pre-rounded to the expected precision by the time this object is constructed. So the rounding in this method is unnecessary. I'll simplify it.

Reply:
I'll also add validation or rounding in the various constructors.

Reply:
This method is responsible for rounding (data is not pre-rounded) when the client does not support new capability.
For this cases at least please do add test cases. (precision=3, values: any)
(In the test you also have a comment why you are not testing certain cases, as they are not possible in real code.)

Reply:
No, this method is not the one that's used for rendering to the client. It's the default toString implementation for SqlTimestamp. In any case, I simplified this whole thing a bit. I added a `roundTo()` method to SqlTimestamp and updated the protocol layer to use that instead and not deal with calling the formatting code directly. I added some tests (and found a minor bug along the way).

Reply:
> No, this method is not the one that's used for rendering to the client.

Perhaps i got mislead because we have similar (but not equal) `formatTimestamp` in SqlTimestamp and Timestamps.
What's the reason?

Reply:
Visibility. SqlTimestamp is in the SPI, but Timestamps is not (and should not be -- it has too many low-level implementation methods)